cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/SQLEngine.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.engine;
+
+import io.cdap.cdap.api.annotation.Beta;
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import io.cdap.cdap.etl.api.PipelineConfigurable;
+import io.cdap.cdap.etl.api.SubmitterLifecycle;
+import io.cdap.cdap.etl.api.batch.BatchContext;
+import io.cdap.cdap.etl.api.join.JoinDefinition;
+
+/**
+ * A SQL Engine can be used to pushdown certain dataset operations.
+ *
+ * SQL Engines are implemented as plugins.
+ */
+@Beta
+public interface SQLEngine extends PipelineConfigurable, SubmitterLifecycle<BatchContext> {
+  String PLUGIN_TYPE = "sqlengine";
+
+  /**
+   * Creates an Output Format Provided that can be used to push records into a SQL Engine.
+   * @param tableName The name of the table to use to store these records.
+   * @return an {@link OutputFormatProvider} instance that can be used to write records to the SQL Engine into the
+   * specified table.
+   */
+  OutputFormatProvider push(String tableName);
+
+  /**
+   * Creates an InputFormatProvider that can be used to pull records from the specified table
+   * @param tableName the name of the table to pull records from
+   * @return AN {@link InputFormatProvider} instance that can be used to read records form the specified table in the
+   * SQL Engine.
+   */
+  InputFormatProvider pull(String tableName);
+
+  /**
+   * Check if this table exists in the SQL Engine
+   * @param tableName the table name
+   * @return boolean specifying if this table exists in the remote engine.
+   */
+  boolean exists(String tableName);
+
+  /**
+   * Check if the supplied Join Definition can be executed in this engine.
+   * @param joinDefinition the join definition to check
+   * @return boolean specifying if this join operation can be executed in the SQl Engine.
+   */
+  boolean canJoin(JoinDefinition joinDefinition);
+
+  /**
+   * Executes the join operation defined by the supplied join Definition and stores the result in the defined table.
+   * @param tableName the name of the table to use to store results
+   * @param joinDefinition the join definition
+   * @return the {@link SQLOperationResult} instance with information about the execution of this task.
+   */
+  SQLOperationResult join(String tableName, JoinDefinition joinDefinition);
```

Comment:
What do you think of making this return a `Future<SQLOperationResult>` instead of having this be a blocking call?

Reply:
I'm wondering what should happen if somebody stops a pipeline in the middle of a join. There will need to be some way to cancel the join.  This could either be done by cancelling the Future, or by refactoring the API a bit so that this returns some operation handle and the caller has explicit methods to poll for status or cancel the operation.

Reply:
Would be good to understand what sort of API the BQ client library uses as well.

Reply:
If we make this operation return a `Future`, then I would argue that **all** operations need to return `Future`.

Let's imagine a scenario where we have sources A, B, C and then 2 joins: A-B and B-C.

During the first join operation, both A and B need to be pushed using an OutputFormatProvider. As this operation is ongoing, both tables will be "locked".

If we then invoke the Join operation, we will need to wait until both A and B are fully pushed before executing the join.

At the same time, in parallel, we need to execute the B - C join. Since B is in the process of getting pushed, we should skip it and just push C, but this is not possible as this is a blocking call.

I'll refactor the API a bit. I'll replace the `exists` method with `getTableStatus` and return one of multiple possible states:

- non_existant (table does not exist and no process to push/generate exists)
- computing (either pushing or waiting for the join operation to complete)
- computed (table exists in the remote dataset)

It will be the responsibility of the SQL Engine to keep track of which tables are in which states, and ensure no "concurrent" operations can happen (e.g. a table cannot be pushed multiple times, and a table cannot be pulled before it has been generated by a job). 

Let me know your thoughts. I'll update the design doc with this.

Reply:
I just realized the cleanup method has a flag passed in, which looks like it would be used to tell the engine to cancel any ongoing joins as well as perform cleanup. Seems like the assumption is the engine will keep track of state and everything. 

I think we should try to take that responsibility away from the engine and have it in the caller. If the engine is responsible for state and concurrency, that sort of logic will most likely need to be replicated in almost the same fashion in every engine implementation.

Took a quick look at the BQ API, it looks something like:

```
    JobInfo jobInfo = JobInfo.newBuilder(jobConf)
      .setJobId(jobId)
      .build();
    Job job = bigQuery.create(jobInfo);
    job.waitFor();
```

Something like this seems to make sense, as the caller would be able to kick off multiple joins and wait for the ones it knows it needs to wait for, as well as being able to cancel with `job.cancel()` and clean it up with something like `job.close()` or `job.cleanup()`, and get statistics with `job.getRowsAffect()` or something like that. This way all the required information can be carried along with each individual job, rather than all being in the engine. The push and pull methods return the outputformat/inputformat that the caller then uses to actually perform the push/pull, so the caller will be responsible for using the result appropriately.

Reply:
I see, good point. I will refactor this and update the DD.